### DISCUSSION TITLE
candle::Tensor vs candle_core::Tensor

### DISCUSSION THREAD

**Discussion by vishpat** [huggingface/candle#1773] [Q&A] [answered]
What is the difference between 
```
candle_core::tensor::Tensor vs candle_core::Tensor
```

**Comment by cirocavani** [comment, marked as answer]
As I can see, they are the same.

https://github.com/huggingface/candle/blob/4fd00b890036ef67391a9cc03f896247d0a75711/candle-core/src/lib.rs#L84

Also, I think `candle::` instead of `candle_core::` is just an artifact of how the workspace is configured.

For example, in `candle-examples`:

https://github.com/huggingface/candle/blob/4fd00b890036ef67391a9cc03f896247d0a75711/candle-examples/Cargo.toml#L14

https://github.com/huggingface/candle/blob/4fd00b890036ef67391a9cc03f896247d0a75711/Cargo.toml#L34

So, when using candle published crates, you will have `candle_core::`.

For example, to convert an example from `candle-examples` to use the crates, you have to rewrite the "uses" of candle.

Something like:

```sh
(your project)
cargo add candle-core@0.4.0
cargo add candle-nn@0.4.0
cargo add candle-transformers@0.4.0

curl -kLo src/main.rs https://github.com/huggingface/candle/raw/6400e1b0a08d594e1448d522a41bddc98c584313/candle-examples/examples/phi/main.rs
sed -i '' 's/candle::/candle_core::/' src/main.rs
```

I did this yesterday to test only the CPU and Metal devices.

Full "decoupling" of the Phi example:

https://github.com/cirocavani/seventeen-candle/blob/8909d6cc03025261b424d6a342845baf978a13b3/phi/README.md
